architecture/replication-layer.md

```diff
@@ -0,0 +1,77 @@
+
+### Leases
+
+Leases are held by a single member of the Raft group (usually the same node as the Raft group leader), and the leaseholder is responsible for receiving `BatchRequests` for all reads and writes related to keys in the range. This has the benefit of improving the speed of reads drastically (which don't need consensus, since a quorum of nodes already agreed with the leaseholder as to the status of the node after the write ).
+
+If there is no Lease holder, any range receiving a request will attempt to become to Lease holder.
+
+Since reads bypass Raft, a new lease holder will, among other things, ascertain that its timestamp cache does not report timestamps smaller than the previous lease holder's (so that it's compatible with reads which may have occurred on the former lease holder). This is accomplished by letting leases enter a stasis period (which is just the expiration minus the maximum clock offset) before the actual expiration of the lease, so that all the next lease holder has to do is set the low water mark of the timestamp cache to its new lease's start time.
+
+As a lease enters its stasis period, no more reads or writes are served, which is undesirable. However, this would only happen in practice if a node became unavailable. In almost all practical situations, no unavailability results since leases are usually long-lived (and/or eagerly extended, which can avoid the stasis period) or proactively transferred away from the lease holder, which can also avoid the stasis period by promising not to serve any further reads until the next lease goes into effect.
+
+#### Co-location with Raft leadership
+
+The range lease is completely separate from Raft leadership, and so without further efforts, Raft leadership and the Range lease might not be held by the same Replica. Since it's expensive to not have these two roles colocated (the lease holder has to forward each proposal to the leader, adding costly RPC round-trips), each lease renewal or transfer also attempts to colocate them. In practice, that means that the mismatch is rare and self-corrects quickly.
+
+### Membership Changes: Rebalance/Repair
+
+Whenever there are changes to a cluster's number of nodes, replicas are rebalanced. What that looks like varies depending on whether the membership change is nodes being added or going offline.
+
+**Nodes added**: The new node Gossips information about itself, indicating that it has space available. The cluster then rebalances some replicas onto the new node based on its zone configs, that control replication with table-level granularity.
+
+**Nodes going offline**: If a member of a Raft group ceases to respond, after 5 minutes, the cluster begins to rebalance by replicating the data the downed node held onto nodes with available space.
+
+#### Rebalancing Replicas
+
+When CockroachDB detects a membership change, ultimately, replicas are moved between ranges.
```

Comment:
replicas are moved between nodes, not ranges

Reply:
Done.